File: src/main/java/com/github/svegon/capi/event/render/BlockRenderTypeCallback.java
package com.github.svegon.capi.event.render;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.block.BlockRenderType;
import net.minecraft.block.BlockState;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

public interface BlockRenderTypeCallback {
    Event<BlockRenderTypeCallback> EVENT = EventFactory.createArrayBacked(BlockRenderTypeCallback.class,
            (state, callback) -> {}, listeners -> (state, callback) -> {
        for (BlockRenderTypeCallback listener : listeners) {
            listener.getBlockRenderType(state, callback);

            if (callback.isCancelled()) {
                return;
            }
        }
    });

    void getBlockRenderType(BlockState state, CallbackInfoReturnable<BlockRenderType> callback);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/render/BlockRenderTypeCallback.kt
package io.github.svegon.capi.event.render

import io.github.svegon.capi.event.render.BlockRenderTypeCallback
import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.block.BlockRenderType
import net.minecraft.block.BlockState
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable
import java.util.function.Function
import kotlin.Array
import kotlin.invoke

interface BlockRenderTypeCallback {
    fun getBlockRenderType(state: BlockState?, callback: CallbackInfoReturnable<BlockRenderType?>?)

    companion object {
        @JvmField
        val EVENT: Event<BlockRenderTypeCallback?> = EventFactory.createArrayBacked(BlockRenderTypeCallback::class.java,
            BlockRenderTypeCallback { state: BlockState?, callback: CallbackInfoReturnable<BlockRenderType?>? -> },
            Function { listeners: Array<BlockRenderTypeCallback?>? ->
                BlockRenderTypeCallback { state: BlockState?, callback: CallbackInfoReturnable<BlockRenderType?>? ->
                    for (listener in listeners) {
                        listener.getBlockRenderType(state, callback)

                        if (callback.isCancelled()) {
                            return@BlockRenderTypeCallback
                        }
                    }
                }
            })
    }
}
